File: 2021-09/2021-9-30 223.矩形面积/computeArea.go
package main

import (
	"fmt"
)

//-2
//-2
//2
//2
//3
//3
//4
//4

func main() {
	ax1 := -2
	ay1 := -2
	ax2 := 2
	ay2 := 2
	bx1 := 2
	by1 := 0
	bx2 := 3
	by2 := 1
	computeArea(ax1, ay1, ax2, ay2, bx1, by1, bx2, by2)
}

func computeArea(ax1 int, ay1 int, ax2 int, ay2 int, bx1 int, by1 int, bx2 int, by2 int) int {
	var conflictArea int
	// 判断两矩形是否重合
	if bx1 > ax2 || bx2 < ax1 || by1 > ay2 || by2 < ay1 {
		conflictArea = 0
	} else {
		// 算重叠的面积
		conflictArea = (min(ax2, bx2) - max(bx1, ax1)) * (min(by2, ay2) - max(by1, ay1))
	}
	// 求出两个矩形的面积不考虑重叠
	area := ((ax2 - ax1) * (ay2 - ay1)) + ((bx2 - bx1) * (by2 - by1))
	fmt.Println(area, conflictArea)
	return area - conflictArea
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 2021/2021-09/2021-9-30 223.矩形面积/ComputeArea.cs
using System;

public class Solution
{
    public int ComputeArea(int ax1, int ay1, int ax2, int ay2, int bx1, int by1, int bx2, int by2)
    {
        // 右上角的点大于左下角的点
        var area1 = (ax2 - ax1) * (ay2 - ay1);
        var area2 = (bx2 - bx1) * (by2 - by1);
        var overlapWidth = Math.Min(ax2, bx2) - Math.Max(ax1, bx1);
        var overlapHeight = Math.Min(ay2, by2) - Math.Max(ay1, by1);
        var overlapArea = Math.Max(overlapWidth, 0) * Math.Max(overlapHeight, 0);
        return area1 + area2 - overlapArea;
    }
}